### PR TITLE
=str add VirtualProcessor

### PR DESCRIPTION
**Pull request by rkuhn**
This is only for reviewing the idea, the name is of course wrong and it also needs tests (and be used of course).

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala [hunk 1]
```diff
@@ -293,6 +293,51 @@ private[akka] object StreamLayout {
...
+  import V._
+  import ReactiveStreamsCompliance._
+
+  private val subscriptionStatus = new AtomicReference[AnyRef]
+  private val terminationStatus = new AtomicReference[Termination]
+
+  override def subscribe(s: Subscriber[_ >: T]): Unit =
+    if (subscriptionStatus.compareAndSet(null, s)) () // wait for onSubscribe
+    else {
+      val sub = subscriptionStatus.getAndSet(s).asInstanceOf[Subscription]
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: this will fail on the second subscribe call with a ClassCastException

Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: Right, this should be turned into a SpecViolation instead.

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala [hunk 2]
```diff
@@ -293,6 +293,51 @@ private[akka] object StreamLayout {
...
+      tryOnSubscribe(s, sub)
+      terminationStatus.getAndSet(Allowed) match {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: this will be null if onSubscribe was first, since it does not set terminationStatus after the successful CAS on subscriptionStatus, and this will enter this branch since the CAS will fail here on subscriptionStatus. Maybe you wanted this to have the initial value of Allowed?

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala (hunk 1 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: It is not a spec violation, just that this VirtualProcessor does not support fanout. This is btw one of the drawbacks of this approach, if you use Sink.publisher to access the features a possibly more powerful Publisher to the left, now it is reduced to single-subscription. I think it is fine, just for the record.

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala (hunk 2 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: No, you just need to ignore the null, since setting it to Allowed in the getAndSet will force the onComplete to fail the CAS and call the onComplete itself

**Comment by drewhk** [comment]
We can test it against the TCK by appending this processor to a real, actor backed identity processor:
 - attach to the left end
 - attach to the right end
 - attach to both ends

**Comment by drewhk** [comment]
I have verified it, and it looks correct (I can present an informal "proof")

TODO:
 - test the mapMaterialized case where the Publisher/Subscriber is accessed and used in the mapper
 - add TCK tests with this appended to left, right, both ends of an IdentityProcessor
 - remove all the Virtual stuff and replace it with this one
 - remove and inline the attach() method in MaterializerSession
 - remove the createIdentityProcessor method in MaterializerSession

**Comment by rkuhn** [comment]
okay, going forward with this now

**Comment by rkuhn** [comment]
Implemented all the things and found a bonus bug as well (more correctly: an inconsistency in how `Keep.left` would completely drop the right computation instead of only discarding its value).

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala [hunk 3]
```diff
@@ -550,6 +620,7 @@ private[stream] abstract class MaterializerSession(val topLevel: StreamLayout.Mo
         case mv: MaterializedValueSource[_] ⇒
           val pub = new MaterializedValuePublisher
           materializedValuePublishers ::= pub
+          materializedValues.put(mv, ())
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: You should run the materialization benchmarks (or add a new one) to see how much impact does this have (given that the default is Keep.left almost always). If it is significantly slower ( > 5%) then I am against this.

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala [hunk 4]
```diff
@@ -293,38 +293,108 @@ private[akka] object StreamLayout {
...
+    tryOnNext(subscriptionStatus.get.asInstanceOf[Subscriber[T]], t)
+  }
+
+  private final class Sub(s: Subscription) extends AtomicLong with Subscription {
+    override def cancel(): Unit = {
+      subscriptionStatus.set(InertSubscriber)
+      s.cancel()
+    }
+    @tailrec
+    override def request(n: Long): Unit = {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: ewww. Why is this needed? There is something frightening about double accounting of requests.

**Comment by drewhk** [comment]
Please also change the materialization of the Identity module to use this instead of Transform(() => Map(identity))

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala (hunk 3 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: This is not optional: without it the tests for accessing the materialized value fail. I don’t expect much impact on real-world benchmarks since this only affects usage of `FlowGraph.Builder.materializedValue`.

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala (hunk 4 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: Without this the TCK fails: it rightfully signals reentrant signaling towards the Subscriber since requesting from onSubscribe may now immediately trigger the asynchronous invocation of onNext. Therefore requests during onSubscribe are delayed by this method.

**Comment by rkuhn** [comment]
Concerning changing the materialization of Identity nodes: is the intention that these will be removed before fusing anyway? In that case I agree.

**Comment by rkuhn** [comment]
@drewhk I have replaced Identity by this VirtualProcessor and then by pure chance I saw the same failure as in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/2953/ locally: ConcatAllImpl suffers from stopping the ActorProcessor too eagerly (when onSubscribe has not yet been received from the substream). Please review the second commit carefully.

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala (hunk 4 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: Ok.

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala (hunk 3 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: Well, let's accept this as a feature then.

**Comment by drewhk** [comment]
> Concerning changing the materialization of Identity nodes: is the intention that these will be removed before fusing anyway? In that case I agree.

These cannot be removed completely trivially in all cases, see

   incomingConneciton.flow.join(Flow[ByteString]) // echo server

I guess it is possible to detect this in the materialization phase, but fusing cannot help here.

**Comment by rkuhn** [comment]
Failure was remaining child actors in TlsSpec—is that one of the typical failure modes or a new one?

@drewhk Please review especially the last three commits since they fix problems that are unrelated to the VirtualProcessor itself, they were racy before and might have come up now due to timing changes.

**Comment by drewhk** [comment]
yes, that is the typical failure.

**Comment by drewhk** [comment]
mergeable?

**Comment by rkuhn** [comment]
Yes, I’d say it is :-)

**Comment by rkuhn** [comment]
Have you seen the improved fix in 5b420c5?

**Comment by drewhk** [comment]
yes, please squash

**Comment by rkuhn** [comment]
It is already squashed: the commits fix different things.
